File: java/src/main/java/br/com/emendes/problems/ZigzagConversion.java
package br.com.emendes.problems;

/**
 * A string "PAYPALISHIRING" é escrita em zigzag pattern em um determinado número de linhas como este:
 * P   A   H   N
 * A P L S I I G
 * Y   i   R
 * E então leia linha por linha: "PAHNAPLSIIGYIR"
 * Escreva o código que pegará uma string e fará essa conversão dado um número de linhas.
 */
public class ZigzagConversion {

  // Boa solução, executou em 2ms.
  public String convert(String s, int numRows) {
    char[] sArray = s.toCharArray();
    int sLength = sArray.length;

    if (numRows == 1 || sLength <= numRows) return s;

    char[] answer = new char[sLength];

    int ci = 0;   // current index de sArray.
    int cr = 0;   // current row
    int ansIndex = 0;   // index do array de resposta.
    int constant = (numRows << 1) - 2;
    int doubleCR;

    while (ci < sLength) { // calcula todos os indices da primeira row.
      answer[ansIndex++] = sArray[ci];
      ci = constant + ci;
    }
    cr++;

    while (cr < numRows - 1) { //calcula da segunda row até a penúltima.
      ci = cr;
      doubleCR = cr << 1;
      while (ci < sLength) {
        answer[ansIndex++] = sArray[ci];
        ci = constant - doubleCR + ci;

        if (ci >= sLength) break; // Após o calculo do próximo index, se ci >= sLength, então saí do loop.

        answer[ansIndex++] = sArray[ci];
        ci = doubleCR + ci;
      }
      cr++;
    }

    ci = cr;
    doubleCR = cr << 1;
    while (ci < sLength) { // Calcula para a última row.
      answer[ansIndex++] = sArray[ci];
      ci = doubleCR + ci;
    }


    return String.valueOf(answer);
  }

}

File: javascript/problems/ZigzagConversion.js
/**
 * A string "PAYPALISHIRING" é escrito no padrão zigzag em um dado número de linhas desse jeito:
 *
 * P   A   H   N
 * A P L S I I G
 * Y   I   R
 *
 * e então lido linha por linha: "PAHNAPLSIIGYIR"
 *
 * Escreva o código que recebe uma string e o número de linhas e converta para o modo zigzag.
 *
 * Restrições:
 * 1 <= s.length <= 1000
 *
 * s consiste de letras (maiúsculas e minúsculas), ',' e '.'.
 *
 * 1 <= numsRows <= 1000
 *
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function (s, numRows) {
  if (numRows === 1) return s;
  let rows = new Array(numRows).fill("");
  let r = 0;
  let inc = 1;
  numRows--;

  for (const element of s) {
    rows[r] += element;
    r += inc;
    if (r === 0 || r >= numRows) {
      inc *= -1;
    }
  }
  return rows.join("");
};
